### ISSUE TITLE
Path.GetDirectoryName converts file:/// to file:/ or file:\

### ISSUE THREAD

**Issue by EgorBo** [dotnet/corefx#30213]
The following code:
```csharp
var assemblyCodeBase = typeof(Program).Assembly.CodeBase;
Console.WriteLine($"assemblyCodeBase={assemblyCodeBase}");
var directoryName = Path.GetDirectoryName(assemblyCodeBase);
Console.WriteLine($"directoryName={directoryName}");
var uri = new Uri(directoryName);
Console.WriteLine($"uri={uri}");
```
works on Windows but throws `System.UriFormatException: Invalid URI: The Authority/Host could not be parsed.` on macOS
It's probably not a good idea to use `Uri` for files but I am more concerned about `Path.GetDirectoryName` removes slashes (see below).

Output for Windows:
```
assemblyCodeBase=file:///C:/Users/egorb/ConsoleApp14/ConsoleApp14.dll
directoryName=file:\C:\Users\egorb\ConsoleApp14
uri=file:///C:/Users/egorb/ConsoleApp14
```
Output for macOS:
```
assemblyCodeBase=file:///Users/egorb/ConsoleApp14/ConsoleApp14.dll
directoryName=file:/Users/egorb/ConsoleApp14

Unhandled Exception: System.UriFormatException: Invalid URI: The Authority/Host could not be parsed.
```
Is it expected behavior?

.NET Core 2.1-300

**Comment by JeremyKuhne** [comment]
> Is it expected behavior?

Yes, it is by design that separators are normalized. `Path` is designed to work efficiently with local paths, and expects all input to be in a local path format.

`Uri` "fixes" bad Uris, but clearly doesn't do it consistently. @davidsh can you comment to the "fixup" behavior expectations?

**Comment by davidsh** [comment]
>`Uri` "fixes" bad Uris, but clearly doesn't do it consistently. @davidsh can you comment to the "fixup" behavior expectations?

@rmkerr Can you please comment on this?

**Comment by rmkerr** [comment]
We do OS specific handling of file paths. Unfortunately it's been a bit of a pain point recently, especially on Unix. I assume what's happening here is that some Unix specific code is interacting poorly with the special handing we have for Windows file paths, and the whole thing is falling over.
> It's probably not a good idea to use `Uri` for files but I am more concerned about `Path.GetDirectoryName` removes slashes (see below).

I would absolutely second that. If possible, avoid using Uri for file paths. Especially on Unix, where we cannot disambiguate relative URIs and local file paths.

**Comment by karelz** [comment]
Triage: Part of group of similar errors. We should in 5.0 decide if to fix it (and fix it), or close it.
